File: 2023/day01/day01.scala
package day01

import locations.Directory.currentDir
import inputs.Input.loadFileSync

@main def part1: Unit =
  println(s"The solution is ${part1(loadInput())}")

@main def part2: Unit =
  println(s"The solution is ${part2(loadInput())}")

def loadInput(): String = loadFileSync(s"$currentDir/../input/day01")

def part1(input: String): String =

  var iter =
    for line <- input.linesIterator yield {
      val digits = line.filter(_.isDigit).map(_.toInt - 48).toList
      val num = digits.head * 10 + digits.last
      num
    }

  iter.sum.toString()

def part2(input: String): String =

  val digitStringsMap = Vector(
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine"
  ).zip(1 to 9).toMap

  val digitLiteralMap = (0 to 9).map { d => (d.toString(), d) }.toMap
  val digitsMap = digitStringsMap ++ digitLiteralMap

  var iter = for line <- input.linesIterator yield {
    val (firstDigit, restOfLine) = digitsMap.keys
      .map(d => (d, line.indexOf(d)))
      .filter { _._2 != -1 }
      .toList
      .sortBy(_(1))
      .headOption
      .map((s, i) => (digitsMap(s), line.substring(i)))
      .get

    val lastDigit = digitsMap.keys
      .map(d => (d, restOfLine.lastIndexOf(d)))
      .filter { _._2 != -1 }
      .toList
      .sortBy(_(1))
      .lastOption
      .map((s, _) => digitsMap(s))
      .get

    firstDigit * 10 + lastDigit
  }

  iter.sum.toString()

File: 2023/haskell1/day01.hs
module Main where

import Data.Char (ord)
import Data.List (find, isInfixOf)
import Data.Map (Map, fromList, (!))

main :: IO ()
main = do
  input <- readFile "../input/day01"
  part1 input
  part2 input

digits :: [Char]
digits = ['0' .. '9']

digitStrings :: [String]
digitStrings = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

digitStringsReverse :: [String]
digitStringsReverse = map reverse digitStrings

digitStringsMap :: Map String Int
digitStringsMap = fromList $ zip digitStrings [1 .. 9]

digitStringsReverseMap :: Map String Int
digitStringsReverseMap = fromList $ zip digitStringsReverse [1 .. 9]

startsWith :: (Eq a) => [a] -> [a] -> Bool
startsWith [] _ = True
startsWith _ [] = False
startsWith (x : xt) (y : yt) = (x == y) && startsWith xt yt

findLiteral :: [Char] -> Int
findLiteral cs@(c : ct) =
  if c `elem` digits
    then ord c - 48
    else findLiteral ct

findLiteralOrString :: [Char] -> Int
findLiteralOrString cs@(c : ct) =
  if c `elem` digits
    then ord c - 48
    else case find (`startsWith` cs) digitStrings of
      Just s -> digitStringsMap ! s
      Nothing -> findLiteralOrString ct

findLastLiteralOrString :: [Char] -> Int
findLastLiteralOrString cs@(c : ct) =
  if c `elem` digits
    then ord c - 48
    else case find (`startsWith` cs) digitStringsReverse of
      Just s -> digitStringsReverseMap ! s
      Nothing -> findLastLiteralOrString ct

findDigits :: [Char] -> Int
findDigits line =
  firstDigit * 10 + lastDigit
  where
    firstDigit = findLiteral line
    lastDigit = findLiteral $ reverse line

findDigitsLiteralOrString :: [Char] -> Int
findDigitsLiteralOrString line =
  firstDigit * 10 + lastDigit
  where
    firstDigit = findLiteralOrString line
    lastDigit = findLastLiteralOrString $ reverse line

part1 :: String -> IO ()
part1 input = do
  print $ sum $ map findDigits $ lines input

part2 :: String -> IO ()
part2 input = do
  print $ sum $ map findDigitsLiteralOrString $ lines input
